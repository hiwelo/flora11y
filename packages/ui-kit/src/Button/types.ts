import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';

export type ButtonProps =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;
