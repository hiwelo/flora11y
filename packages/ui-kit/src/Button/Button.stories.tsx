import { text } from '@storybook/addon-knobs';
import React from 'react';
import Button from './Button';

export default {
  component: Button,
  title: 'Button',
};

export const overview: React.FC = () => (
  <Button>{text('Content', 'Hello World!')}</Button>
);

export const linkMarkup: React.FC = () => (
  <Button href={text('Hyperlink', '')}>
    {text('Content', 'Hello World!')}
  </Button>
);
