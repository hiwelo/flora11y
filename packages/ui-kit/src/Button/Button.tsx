import React from 'react';

import StyledButton from './components/StyledButton';
import { ButtonProps } from './types';

const Button: React.FC<ButtonProps> = ({
  children,
  disabled,
  focusable,
  href,
  ...rest
}) => {
  const isLink = !!href && !disabled;
  const isButton = !isLink;
  const isDisabled = isButton && disabled;

  return (
    <StyledButton
      Element={isLink ? 'a' : 'button'}
      aria-disabled={disabled}
      disabled={(isDisabled && !focusable) || undefined}
      href={isLink ? href : undefined}
      type={isButton ? 'button' : undefined}
      {...rest}
    >
      {children}
    </StyledButton>
  );
};

export default Button;
