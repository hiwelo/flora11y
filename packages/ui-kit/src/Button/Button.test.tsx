import { render, screen } from '@testing-library/react';
import React from 'react';
import { create, act } from 'react-test-renderer';

import Button from '.';

const testElements = (
  <>
    <Button data-testid="default">Hello World!</Button>
    <Button data-testid="link" href="#">
      Hello World!
    </Button>
    <Button data-testid="disabledButton" disabled>
      Hello World!
    </Button>
    <Button data-testid="disabledLink" disabled href="#">
      Hello World!
    </Button>
    <Button data-testid="focusableDisabledButton" disabled focusable>
      Hello World!
    </Button>
  </>
);

describe('<Button />', () => {
  it('should adapt the markup based on the provided props', () => {
    expect.assertions(2);
    render(testElements);

    expect(screen.queryByTestId('default')?.tagName).toBe('BUTTON');
    expect(screen.queryByTestId('link')?.tagName).toBe('A');
  });

  it('should always render a button when disabled, even with an URL target', () => {
    expect.assertions(2);
    render(testElements);

    expect(screen.queryByTestId('disabledButton')?.tagName).toBe('BUTTON');
    expect(screen.queryByTestId('disabledLink')?.tagName).toBe('BUTTON');
  });

  it('should be render a focusable disabled button when requested', () => {
    expect.assertions(2);
    render(testElements);

    const disabledFocusableButton = screen.queryByTestId(
      'focusableDisabledButton',
    );
    const disabledButton = screen.queryByTestId('disabledButton');

    act(() => {
      if (!disabledFocusableButton) return;

      disabledFocusableButton.focus();
    });

    expect(document.activeElement).toBe(disabledFocusableButton);

    act(() => {
      if (!disabledButton) return;

      disabledButton.focus();
    });

    expect(document.activeElement).not.toBe(disabledButton);
  });

  it('should match snapshot', () => {
    expect.hasAssertions();
    const container = create(testElements);

    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(container.toJSON()).toMatchSnapshot();
  });
});
