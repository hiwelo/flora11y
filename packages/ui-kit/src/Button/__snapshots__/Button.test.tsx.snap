// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
Array [
  <button
    className="sc-AxjAm jlznlR"
    data-testid="default"
    type="button"
  >
    Hello World!
  </button>,
  <a
    className="sc-AxjAm jlznlR"
    data-testid="link"
    href="#"
  >
    Hello World!
  </a>,
  <button
    aria-disabled={true}
    className="sc-AxjAm jlznlR"
    data-testid="disabledButton"
    disabled={true}
    type="button"
  >
    Hello World!
  </button>,
  <button
    aria-disabled={true}
    className="sc-AxjAm jlznlR"
    data-testid="disabledLink"
    disabled={true}
    type="button"
  >
    Hello World!
  </button>,
  <button
    aria-disabled={true}
    className="sc-AxjAm jlznlR"
    data-testid="focusableDisabledButton"
    type="button"
  >
    Hello World!
  </button>,
]
`;
